### PR TITLE
Update WR (various optimizations, gradient improvements).

### PR DESCRIPTION
**Pull request by glennw**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16200)

### CODE REVIEW

File: components/canvas/webgl_paint_thread.rs [hunk 1]
```diff
@@ -250,22 +250,36 @@ impl WebGLPaintThread {
...
+                                     format: webrender_traits::ImageFormat::RGBA8,
+                                     offset: 0,
+                                     is_opaque: false,
+                                 };
+                let data = webrender_traits::ImageData::Raw(Arc::new(pixels));
+
+                match *image_key {
+                    Some(image_key) => {
+                        // TODO: This shouldn't be a common path, but try to avoid
+                        // the spurious clone().
```
Comment (jdm) [inline components/canvas/webgl_paint_thread.rs]: No such thing any more.

File: components/compositing/compositor.rs [hunk 2]
```diff
@@ -76,9 +76,9 @@ trait ConvertScrollRootIdFromWebRender {
     fn from_webrender(&self) -> ScrollRootId;
 }
 
-impl ConvertScrollRootIdFromWebRender for usize {
+impl ConvertScrollRootIdFromWebRender for u64 {
     fn from_webrender(&self) -> ScrollRootId {
-        ScrollRootId(*self)
+        ScrollRootId(*self as usize)
```
Comment (jdm) [inline components/compositing/compositor.rs]: Maybe update the inner type in ScrollRootId?

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by glennw** [comment]
@jdm Fixed the comment.

I left the internal type as usize because it relies on an AtomicUsize. The AtomicU64 type is marked as unstable - I thought it would be best to not introduce a requirement for the unstable AtomicU64 (it's also not clear to me if 32bit android guarantees AtomicU64 is available). Switching the type to u64 but keeping the atomic as usize introduces more casts than the current PR, I think. Thoughts?

**Comment by jdm** [comment]
That seems like a reasonable choice.

File: components/canvas/canvas_paint_thread.rs [hunk 3]
```diff
@@ -558,20 +558,35 @@ impl<'a> CanvasPaintThread<'a> {
...
-                                                stride: None,
-                                                format: webrender_traits::ImageFormat::RGBA8,
-                                                offset: 0,
-                                                is_opaque: false,
-                                            },
-                                            element.into(),
-                                            None);
+
+            let descriptor = webrender_traits::ImageDescriptor {
+                                width: size.width as u32,
```
Comment (emilio) [inline components/canvas/canvas_paint_thread.rs]: nit: Indentation looks weird here. Either indent membersfour spaces from `webrender_traits`, or move it to the left so the members are aligned with `descriptor`

File: components/compositing/compositor.rs (hunk 2 above)
Comment (emilio) [inline components/compositing/compositor.rs]: The problem with this is that is lossy in 32-bit platforms... But probably that's ok assuming we only pass `usize`s to WR. Perhaps leave a comment or something?

**Review by emilio — COMMENTED** [review]
Looks good to me with nits, thanks!

**Comment by glennw** [comment]
@emilio Rebased and fixed nits.

**Pull request merged**